### PR TITLE
Iamse 1169 external authentication

### PR DESCRIPTION
**Pull request by scalding**
Add the ability to use external authentication in Grouper

### CODE REVIEW

File: grouper-container/grouper-uiDaemon-container/pom.xml [hunk 1]
```diff
@@ -28,47 +28,52 @@
     <dependency>
       <groupId>${project.groupId}</groupId>
       <artifactId>grouper-ui</artifactId>
-      <version>${grouper.version}</version>
```
Comment (scalding) [inline grouper-container/grouper-uiDaemon-container/pom.xml:31]: Review the need for this change

File: grouper-misc/grouper-authentication/conf/grouper.external.authentication.example.properties [hunk 2]
```diff
@@ -0,0 +1,96 @@
+# these are properties to configure PAC4J with CAS
```
Comment (scalding) [inline grouper-misc/grouper-authentication/conf/grouper.external.authentication.example.properties:1]: Make sure the sample properties are up to date

File: grouper-misc/grouper-authentication/src/main/java/edu/internet2/middleware/grouper/authentication/ConfigUtils.java [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+import java.lang.reflect.Field;
+import java.lang.reflect.Type;
+import java.time.Period;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+
+public class ConfigUtils {
```
Comment (scalding) [inline grouper-misc/grouper-authentication/src/main/java/edu/internet2/middleware/grouper/authentication/ConfigUtils.java:19]: add javadocs

File: grouper-misc/grouper-authentication/src/main/java/edu/internet2/middleware/grouper/authentication/Pac4jConfigFactory.java [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+import edu.internet2.middleware.grouper.authentication.config.ClientProviders;
+import edu.internet2.middleware.grouper.ui.util.GrouperUiConfig;
+import org.apache.log4j.Logger;
+import org.pac4j.core.client.Client;
+import org.pac4j.core.client.Clients;
+import org.pac4j.core.config.Config;
+import org.pac4j.core.config.ConfigFactory;
+import org.pac4j.core.matching.matcher.PathMatcher;
+
+public class Pac4jConfigFactory implements ConfigFactory {
```
Comment (scalding) [inline grouper-misc/grouper-authentication/src/main/java/edu/internet2/middleware/grouper/authentication/Pac4jConfigFactory.java:14]: add javadocs

File: grouper-misc/grouper-authentication/src/main/java/edu/internet2/middleware/grouper/authentication/Pac4jServletContainerInitializer.java [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.servlet.FilterRegistration;
+import javax.servlet.ServletContainerInitializer;
+import javax.servlet.ServletContext;
+import javax.servlet.ServletException;
+import java.util.Set;
+
+public class Pac4jServletContainerInitializer implements ServletContainerInitializer {
```
Comment (scalding) [inline grouper-misc/grouper-authentication/src/main/java/edu/internet2/middleware/grouper/authentication/Pac4jServletContainerInitializer.java:15]: add javadoc

File: grouper-misc/grouper-authentication/src/main/java/edu/internet2/middleware/grouper/authentication/config/CasClientProvider.java [hunk 6]
```diff
@@ -0,0 +1,23 @@
+package edu.internet2.middleware.grouper.authentication.config;
+
+import edu.internet2.middleware.grouper.authentication.ConfigUtils;
+import org.pac4j.cas.client.CasClient;
+import org.pac4j.cas.config.CasConfiguration;
+import org.pac4j.core.client.Client;
+
+public class CasClientProvider implements ClientProvider {
```
Comment (scalding) [inline grouper-misc/grouper-authentication/src/main/java/edu/internet2/middleware/grouper/authentication/config/CasClientProvider.java:8]: add javadoc

File: grouper-misc/grouper-authentication/src/main/java/edu/internet2/middleware/grouper/authentication/config/ClientProvider.java [hunk 7]
```diff
@@ -0,0 +1,8 @@
+package edu.internet2.middleware.grouper.authentication.config;
+
+import org.pac4j.core.client.Client;
+
+public interface ClientProvider {
```
Comment (scalding) [inline grouper-misc/grouper-authentication/src/main/java/edu/internet2/middleware/grouper/authentication/config/ClientProvider.java:5]: add javadoc

File: grouper-misc/grouper-authentication/src/main/java/edu/internet2/middleware/grouper/authentication/config/ClientProviders.java [hunk 8]
```diff
@@ -0,0 +1,23 @@
+package edu.internet2.middleware.grouper.authentication.config;
+
+import java.util.Locale;
+
+public enum ClientProviders {
```
Comment (scalding) [inline grouper-misc/grouper-authentication/src/main/java/edu/internet2/middleware/grouper/authentication/config/ClientProviders.java:5]: add javadoc

File: grouper-misc/grouper-authentication/src/main/java/edu/internet2/middleware/grouper/authentication/config/OidcClientProvider.java [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+
+import edu.internet2.middleware.grouper.authentication.ConfigUtils;
+import edu.internet2.middleware.grouper.authentication.oidc.config.ClaimAsUsernameOidcConfiguration;
+import edu.internet2.middleware.grouper.ui.util.GrouperUiConfig;
+import org.apache.log4j.Logger;
+import org.pac4j.core.client.Client;
+import org.pac4j.oidc.client.OidcClient;
+import org.pac4j.oidc.config.OidcConfiguration;
+
+public class OidcClientProvider implements ClientProvider {
```
Comment (scalding) [inline grouper-misc/grouper-authentication/src/main/java/edu/internet2/middleware/grouper/authentication/config/OidcClientProvider.java:12]: add javadoc

File: grouper-misc/grouper-authentication/src/main/java/edu/internet2/middleware/grouper/authentication/config/SAML2ClientProvider.java [hunk 10]
```diff
@@ -0,0 +1,24 @@
+package edu.internet2.middleware.grouper.authentication.config;
+
+import edu.internet2.middleware.grouper.authentication.ConfigUtils;
+import org.pac4j.core.client.Client;
+import org.pac4j.saml.client.SAML2Client;
+import org.pac4j.saml.config.SAML2Configuration;
+
+public class SAML2ClientProvider implements ClientProvider {
```
Comment (scalding) [inline grouper-misc/grouper-authentication/src/main/java/edu/internet2/middleware/grouper/authentication/config/SAML2ClientProvider.java:8]: add javadoc

File: grouper-misc/grouper-authentication/src/main/java/edu/internet2/middleware/grouper/authentication/oidc/client/ClaimAsUsernameOidcClient.java [hunk 11]
```diff
@@ -0,0 +1,18 @@
+package edu.internet2.middleware.grouper.authentication.oidc.client;
+
+import edu.internet2.middleware.grouper.authentication.oidc.config.ClaimAsUsernameOidcConfiguration;
+import edu.internet2.middleware.grouper.authentication.oidc.profile.ClaimAsUsernameProfileCreator;
+import org.pac4j.oidc.client.OidcClient;
+
+public class ClaimAsUsernameOidcClient extends OidcClient<ClaimAsUsernameOidcConfiguration> {
```
Comment (scalding) [inline grouper-misc/grouper-authentication/src/main/java/edu/internet2/middleware/grouper/authentication/oidc/client/ClaimAsUsernameOidcClient.java:7]: add javadoc

File: grouper-misc/grouper-authentication/src/main/java/edu/internet2/middleware/grouper/authentication/oidc/config/ClaimAsUsernameOidcConfiguration.java [hunk 12]
```diff
@@ -0,0 +1,15 @@
+package edu.internet2.middleware.grouper.authentication.oidc.config;
+
+import org.pac4j.oidc.config.OidcConfiguration;
+
+public class ClaimAsUsernameOidcConfiguration extends OidcConfiguration {
```
Comment (scalding) [inline grouper-misc/grouper-authentication/src/main/java/edu/internet2/middleware/grouper/authentication/oidc/config/ClaimAsUsernameOidcConfiguration.java:5]: add javadoc

File: grouper-misc/grouper-authentication/src/main/java/edu/internet2/middleware/grouper/authentication/oidc/profile/ClaimAsUsernameProfile.java [hunk 13]
```diff
@@ -0,0 +1,16 @@
+package edu.internet2.middleware.grouper.authentication.oidc.profile;
+
+import org.pac4j.oidc.profile.OidcProfile;
+
+public class ClaimAsUsernameProfile extends OidcProfile {
```
Comment (scalding) [inline grouper-misc/grouper-authentication/src/main/java/edu/internet2/middleware/grouper/authentication/oidc/profile/ClaimAsUsernameProfile.java:5]: add javadoc

File: grouper-misc/grouper-authentication/src/main/java/edu/internet2/middleware/grouper/authentication/oidc/profile/ClaimAsUsernameProfileCreator.java [hunk 14]
```diff
@@ -0,0 +1,22 @@
+package edu.internet2.middleware.grouper.authentication.oidc.profile;
+
+import edu.internet2.middleware.grouper.authentication.oidc.config.ClaimAsUsernameOidcConfiguration;
+import org.pac4j.core.util.CommonHelper;
+import org.pac4j.oidc.client.OidcClient;
+import org.pac4j.oidc.config.OidcConfiguration;
+import org.pac4j.oidc.profile.creator.OidcProfileCreator;
+
+public class ClaimAsUsernameProfileCreator extends OidcProfileCreator<ClaimAsUsernameProfile> {
```
Comment (scalding) [inline grouper-misc/grouper-authentication/src/main/java/edu/internet2/middleware/grouper/authentication/oidc/profile/ClaimAsUsernameProfileCreator.java:9]: add javadoc

File: grouper-misc/grouper-authentication/src/main/java/edu/internet2/middleware/grouper/authentication/oidc/profile/ClaimAsUsernameProfileDefinition.java [hunk 15]
```diff
@@ -0,0 +1,10 @@
+package edu.internet2.middleware.grouper.authentication.oidc.profile;
+
+import org.pac4j.oidc.profile.OidcProfileDefinition;
+
+public class ClaimAsUsernameProfileDefinition extends OidcProfileDefinition<ClaimAsUsernameProfile> {
```
Comment (scalding) [inline grouper-misc/grouper-authentication/src/main/java/edu/internet2/middleware/grouper/authentication/oidc/profile/ClaimAsUsernameProfileDefinition.java:5]: add javadoc

**Comment by mchyzer** [comment]
You asked me to leave a note with the reminder about dynamic configuration so grouper doesnt need to be bounced to have changed take effect :)

**Comment by scalding** [comment]
closing in favor of something else

**Pull request closed without merging**